File: src/CinemaVerse.Services/Implementations/Admin/AdminDashboardService.cs
using CinemaVerse.Data.Repositories;
using CinemaVerse.Services.Interfaces.Admin;
using Microsoft.Extensions.Logging;

namespace CinemaVerse.Services.Implementations.Admin
{
    public class AdminDashboardService : IAdminDashboard
    {
        private readonly IUnitOfWork _unitOfWork;
        private readonly ILogger<AdminDashboardService> _logger;

        public AdminDashboardService(IUnitOfWork unitOfWork, ILogger<AdminDashboardService> logger)
        {
            _unitOfWork = unitOfWork;
            _logger = logger;
        }
        public async Task<decimal> CalculateOccupancyRate()
        {
            try
            {
                _logger.LogInformation("Calculating occupancy rate for the admin dashboard.");
                var occupancyRate = await _unitOfWork.Dashboard.CalculateOccupancyRate();
                _logger.LogInformation("Occupancy rate calculated: {OccupancyRate}%", occupancyRate);
                return occupancyRate;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error calculating occupancy rate for the admin dashboard.");
                throw;
            }
        }

        public async Task<decimal> GetActiveUsers()
        {
            try
            {
                _logger.LogInformation("Calculating active users for the admin dashboard.");
                var activeUsers = await _unitOfWork.Dashboard.GetActiveUsers();
                _logger.LogInformation("Active users calculated: {ActiveUsersCount}", activeUsers);
                return activeUsers;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error calculating active users for the admin dashboard.");
                throw;
            }
        }

        public async Task<Dictionary<string, decimal>> GetMonthlyRevenue()
        {
            try
            {
                _logger.LogInformation("Calculating monthly revenue for the admin dashboard.");
                var monthlyRevenue = await _unitOfWork.Dashboard.GetMonthlyRevenue();
                _logger.LogInformation("Monthly revenue calculated.");
                return monthlyRevenue;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error calculating monthly revenue for the admin dashboard.");
                throw;
            }
        }

        public async Task<decimal> GetTotalBookings()
        {
            try
            {
                _logger.LogInformation("Calculating total bookings for the admin dashboard.");
                var totalBookings = await _unitOfWork.Dashboard.GetTotalBookings();
                _logger.LogInformation("Total bookings calculated: {TotalBookingsCount}", totalBookings);
                return totalBookings;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error calculating total bookings for the admin dashboard.");
                throw;
            }
        }

        public async Task<decimal> GetTotalRevenue()
        {
            try
            {
                _logger.LogInformation("Calculating total revenue for the admin dashboard.");
                var totalRevenue = await _unitOfWork.Dashboard.GetTotalRevenue();
                _logger.LogInformation("Total revenue calculated: {TotalRevenue}", totalRevenue);
                return totalRevenue;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error calculating total revenue for the admin dashboard.");
                throw;
            }
        }

        public async Task<Dictionary<string, decimal>> GetWeeklyBookings()
        {
            try
            {
                _logger.LogInformation("Calculating weekly bookings for the admin dashboard.");
                var weeklyBookings = await _unitOfWork.Dashboard.GetWeeklyBookings();
                _logger.LogInformation("Weekly bookings calculated.");
                return weeklyBookings;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error calculating weekly bookings for the admin dashboard.");
                throw;
            }
        }
    }
}

File: src/frontend/src/app/features/admin/features/dashboard/data-access/admin-dashboard.service.ts
import { Injectable } from '@angular/core';
import { catchError, forkJoin, map, Observable, of } from 'rxjs';
import { ApiClientService } from '../../../../../core/http/api-client.service';

export interface KpiData {
  title: string;
  value: string;
  icon: string;
  subtitle?: string;
  trend?: string;
  trendType?: 'up' | 'down' | 'neutral';
}

export interface ChartSeriesData {
  labels: string[];
  data: number[];
}

export interface AdminDashboardSummaryResponse {
  totalRevenue?: number;
  totalBookings?: number;
  activeUsers?: number;
  occupancyRate?: number;
}

type TimeseriesMetrics = Record<string, number>;
type ChartSeriesApiResponse = ChartSeriesData | TimeseriesMetrics;

@Injectable({ providedIn: 'root' })
export class AdminDashboardService {
  constructor(private readonly apiClient: ApiClientService) {}

  getDashboardKpis(): Observable<KpiData[]> {
    return forkJoin({
      totalRevenue: this.getTotalRevenue().pipe(catchError(() => of(0))),
      totalBookings: this.getTotalBookings().pipe(catchError(() => of(0))),
      activeUsers: this.getActiveUsers().pipe(catchError(() => of(0))),
      occupancyRate: this.getOccupancyRate().pipe(catchError(() => of(0))),
    }).pipe(
      map(({ totalRevenue, totalBookings, activeUsers, occupancyRate }): KpiData[] => [
        {
          title: 'Total Revenue',
          value: this.formatCurrency(totalRevenue),
          icon: 'payments',
          subtitle: 'Admin API · /api/admin/dashboard/total-revenue',
        },
        {
          title: 'Total Bookings',
          value: this.formatCount(totalBookings),
          icon: 'confirmation_number',
          subtitle: 'Admin API · /api/admin/dashboard/total-bookings',
        },
        {
          title: 'Active Users',
          value: this.formatCount(activeUsers),
          icon: 'group',
          subtitle: 'Admin API · /api/admin/dashboard/active-users',
        },
        {
          title: 'Occupancy Rate',
          value: this.formatPercentage(occupancyRate),
          icon: 'chair',
          subtitle: 'Admin API · /api/admin/dashboard/occupancy-rate',
        },
      ]),
    );
  }

  getDashboardSummary(): Observable<AdminDashboardSummaryResponse> {
    return this.apiClient.get<AdminDashboardSummaryResponse>('/api/admin/dashboard');
  }

  getTotalRevenue(): Observable<number> {
    return this.apiClient.get<number>('/api/admin/dashboard/total-revenue');
  }

  getTotalBookings(): Observable<number> {
    return this.apiClient.get<number>('/api/admin/dashboard/total-bookings');
  }

  getActiveUsers(): Observable<number> {
    return this.apiClient.get<number>('/api/admin/dashboard/active-users');
  }

  getOccupancyRate(): Observable<number> {
    return this.apiClient.get<number>('/api/admin/dashboard/occupancy-rate');
  }

  getMonthlyRevenue(): Observable<ChartSeriesData> {
    return this.apiClient
      .get<ChartSeriesApiResponse>('/api/admin/dashboard/monthly-revenue')
      .pipe(map((response) => this.toChartSeriesData(response)));
  }

  getWeeklyBookings(): Observable<ChartSeriesData> {
    return this.apiClient
      .get<ChartSeriesApiResponse>('/api/admin/dashboard/weekly-bookings')
      .pipe(map((response) => this.toChartSeriesData(response)));
  }

  private formatCurrency(value: number): string {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      maximumFractionDigits: 0,
    }).format(value);
  }

  private formatCount(value: number): string {
    return new Intl.NumberFormat('en-US', { maximumFractionDigits: 0 }).format(value);
  }

  private formatPercentage(value: number): string {
    return `${value.toFixed(1)}%`;
  }

  private toChartSeriesData(response: ChartSeriesApiResponse): ChartSeriesData {
    if (this.isChartSeriesData(response)) {
      return response;
    }

    const entries = Object.entries(response);
    return {
      labels: entries.map(([label]) => label),
      data: entries.map(([, value]) => value),
    };
  }

  private isChartSeriesData(value: ChartSeriesApiResponse): value is ChartSeriesData {
    if (typeof value !== 'object' || value === null) {
      return false;
    }

    const candidate = value as Partial<ChartSeriesData>;
    return Array.isArray(candidate.labels) && Array.isArray(candidate.data);
  }
}
